modules/transport-netty4/src/internalClusterTest/java/org/elasticsearch/http/netty4/Netty4HttpRequestStreamIT.java

```diff
@@ -0,0 +1,124 @@
+import java.util.function.Supplier;
+
+public class Netty4HttpRequestStreamIT extends ESNetty4IntegTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.concatLists(List.of(RequestContentStreamPlugin.class), super.nodePlugins());
+    }
+
+    @Override
+    protected boolean addMockHttpTransport() {
+        return false; // enable http
+    }
+
+    public void testBasicStream() throws IOException {
+        var totalBytes = 1024 * 1024;
+        var request = new Request("POST", RequestContentStreamPlugin.ROUTE);
+        request.setEntity(new ByteArrayEntity(randomByteArrayOfLength(totalBytes), ContentType.APPLICATION_JSON));
+
+        var respose = getRestClient().performRequest(request);
+        assertEquals(200, respose.getStatusLine().getStatusCode());
+        var gotTotalBytes = new BytesArray(respose.getEntity().getContent().readAllBytes()).utf8ToString();
+        assertEquals("" + totalBytes, gotTotalBytes);
+    }
+
+    public static class RequestContentStreamPlugin extends Plugin implements ActionPlugin {
+
+        static final String ROUTE = "/_test/request-stream/basic";
+        private static final Logger LOGGER = LogManager.getLogger("StreamRestHandler");
+
+        @Override
+        public Collection<RestHandler> getRestHandlers(
+            Settings settings,
+            NamedWriteableRegistry namedWriteableRegistry,
+            RestController restController,
+            ClusterSettings clusterSettings,
+            IndexScopedSettings indexScopedSettings,
+            SettingsFilter settingsFilter,
+            IndexNameExpressionResolver indexNameExpressionResolver,
+            Supplier<DiscoveryNodes> nodesInCluster,
+            Predicate<NodeFeature> clusterSupportsFeature
+        ) {
+            return List.of(new BaseRestHandler() {
+                @Override
+                public String getName() {
+                    return ROUTE;
+                }
+
+                @Override
+                public List<Route> routes() {
+                    return List.of(new Route(RestRequest.Method.POST, ROUTE));
+                }
+
+                @Override
+                protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) {
+                    return channel -> {
+
+                        // Example of handling streamed content
+
+                        var contentStream = request.contentStream();
+                        var totalReceivedBytes = new AtomicInteger();
+
+                        Consumer<HttpContent.Chunk> contentConsumer = (chunk) -> {
+                            LOGGER.info("got next chunk with size {}", chunk.bytes().length());
+                            totalReceivedBytes.addAndGet(chunk.bytes().length());
+                            if (chunk.isLast() == false) {
+                                contentStream.request(1024 * 10); // ask for 10kb, will round up to 16kb
```

Comment:
Could we have another test which shows the backpressure mechanism working (i.e. a REST handler which stops processing chunks, then we check that the client stops sending chunks, then we release the block on the server side and watch the rest of the request go through).

Reply:
Added back pressure test `testClientBackpressure` in `Netty4IncrementalRequestHandlingIT`